jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
+                // drop any references to the corresponding subscriber.
+                this.demand.set(NO_MORE_DEMAND);
+                // TODO: HttpChannelState does not satisfy the contract of Content.Source "If read() has returned a last chunk, this is a no operation."
+                if (finalSignal != FinalSignal.COMPLETE)
+                    this.content.fail(reason);
+                this.content = null;
+                try
                 {
-                    try (AutoLock ignored = lock.lock())
-                    {
-                        // Restore the demand decremented above.
-                        ++demand;
-                        stalled = true;
-                    }
-                    content.demand(this::process);
-                    return;
+                    if (finalSignal == FinalSignal.COMPLETE)
+                        this.subscriber.onComplete();
+                    if (finalSignal == FinalSignal.ERROR)
+                        this.subscriber.onError(reason);
+                    // do nothing on FinalSignal.SUPPRESS
                 }
-
-                if (Content.Chunk.isFailure(chunk))
+                catch (Throwable err)
                 {
-                    terminate();
-                    if (!chunk.isLast())
-                        content.fail(chunk.getFailure());
-                    subscriber.onError(chunk.getFailure());
-                    return;
+                    LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
                 }
+                this.subscriber = null;
+                return Action.SUCCEEDED;
+            }
+
+            Content.Chunk chunk = content.read();
+
+            if (chunk == null)
+            {
+                content.demand(this::iterate);
+                return Action.IDLE;
+            }
 
-                subscriber.onNext(chunk);
+            if (Content.Chunk.isFailure(chunk))
+            {
+                cancel(chunk.getFailure(), FinalSignal.ERROR);
                 chunk.release();
+                return Action.IDLE;
+            }
 
-                if (chunk.isLast())
-                {
-                    terminate();
-                    // Reactive Stream specification rule 2.9 allows Publishers to call onComplete()
-                    // even without demand, and Subscribers must be prepared to handle this case.
-                    subscriber.onComplete();
-                    return;
-                }
+            try
+            {
+                this.subscriber.onNext(chunk);
+            }
+            catch (Throwable err)
+            {
+                cancel(err, FinalSignal.SUPPRESS);
+                LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+            }
+            chunk.release();
+
+            if (chunk.isLast())
+            {
+                cancel(new CompletionException("Source.Content read fully", null), FinalSignal.COMPLETE);
```

Comment:
creating exception to  record a normal completion is  not efficient

Reply:
@gregw replaced with static exception. Should we do the same LastWillSubscription.cancel() because previously Subscription.cancel() didn't not create Exception?